### ISSUE TITLE
[Feature][DTensor] Manage additional `_padded_local_tensor` attribute

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#113045]
Currently, the DTensor tensor subclass manages a `_local_tensor` attribute that represents the local tensor on the given rank. For efficient all-gather/reduce-scatter, we prefer to have a padded local tensor since those collectives require even sizes across ranks, where the padding is determined from the global shape and placements. (E.g., padding is only required on tensor dimensions that have a `Shard` placement.)

The request is for DTensor to additionally manage a `_padded_local_tensor` attribute:
- If a DTensor does not require any padding, then `_local_tensor is _padded_local_tensor`.
- Otherwise, `_local_tensor` is a view into `_padded_local_tensor` with the padding stripped.
- `_local_tensor` is used for computation; however, `_padded_local_tensor` can be used for communication.

---

If DTensor supports, this may solve an important question for per-parameter-sharding FSDP: how to handle `nn.Module._apply()` calls. Without this feature:
- Assuming `torch.Tensor` from `named_parameters()`, if a user calls `module.cpu()` when `module` is on CUDA, then FSDP breaks its invariant that the sharded parameters are views into padded data, and FSDP no longer has the correct references to the sharded parameters.
- Assuming DTensor from `named_parameters()`, if a user calls `module.cpu()` when `module` is on CUDA, then FSDP similarly loses references to the sharded parameters (since subclasses do not use `.data = ...`).

With this feature and having FSDP return DTensor from `named_parameters()`, we can do the following: DTensor defines `.to()` semantics, moving `_padded_local_tensor` and then re-viewing `_local_tensor` into it, and FSDP always retrieves the `_padded_local_tensor` attribute from each DTensor itself instead managing it in FSDP.

**Comment by awgu** [comment]
@wanchaol Two questions if we go with the 'unpad-when-needed' approach:
- Will the `view` op to unpad be differentiable? Differentiable `view` would incur an extra copy from `ViewBackward` to pad gradients.
- What is the implication on Dynamo/AOTAutograd?

**Comment by eellison** [comment]
> @eellison your intern did a padded tensor for CUDAGraphs, I wonder if it can be used here.

Yes, this would be ideal. cc @BoyuanFeng. The largest difficulty imo is getting complete coverage of padding registrations for custom operators. And I think there are a few other remaining TODOs that we didn't finish for demonstration. But we have at least something to build off of that showed good results in limited testing.
